Provide ability to create Gists as part of testing submitted package

gep13 opened the pull request:
- This is an in progress PR for implementing what we discussed, i.e. getting the Creation of Gists stubbed out
- I will continue to push new work into this branch
- Feel free to discuss as it is developing

**TODO**
- [x] Add all tasks required for package moderation
- [x] Add all messages between tasks
- [x] Document the interaction between tasks and messages
- [x] Add configuration for Gist Username and Password
- [x] Add ability to Create Gist

gep13 commented:
@ferventcoder just did my first partial commit of changes to a file, i.e. splitting changes of one file into two commits.  Very cool that you can do this! :-)

ferventcoder commented:
No emojis in commit messages.

gep13 commented:
>@ferventcoder said...
No emojis in commit messages.

:cry: I have removed them :smile_cat:

gep13 commented:
@ferventcoder sanity check please...

This is what I have now setup

| Subscriptions          | Task                          | Publishes              |
|------------------------|-------------------------------|------------------------|
| None                   | StartupTask                   | StartupMessage         |
| StartupMessage         | CheckForSubmittedPackagesTask | ModeratePackageMessage |
| ModeratePackageMessage | ModeratePackageTask           | CreateGistMessage      |
| CreateGistMessage      | CreateGistTask                | None                   |

Make sense?  Anything obvious missing?

**NOTE:** A very cool site for creating markdown tables: http://www.tablesgenerator.com/markdown_tables

ferventcoder commented:
It's close, but the problem is you are directing flow based on the name of the messages. It doesn't allow inserting tasks in between. For instance, look at CheckForSubmittedPackages task. What kind of message does it send? It should send something related to itself, not related to the next thing in line.  Like submitted package message for each submitted package ready for test it finds.

gep13 commented:
Nope, that makes perfect sense.  So based on that suggestion, am I right in saying something more like this is required?

| Subscriptions           | Task                          | Schedule                    | Publishes               |
|-------------------------|-------------------------------|-----------------------------|-------------------------|
| None                    | StartupTask                   | Timer (immediately stopped) | StartupMessage          |
| StartupMessage          | CheckForSubmittedPackagesTask | Timer (15 minutes)          | PackageSubmittedMessage |
| PackageSubmittedMessage | ModeratePackageTask           | None                        | PackageModeratedMessage |
| PackageModeratedMessage | CreateGistTask                | None                        | GistCreatedMessage      |
| GistCreatedMessage      | UpdateWebsiteInformationTask  | None                        | WebsiteUpdatedMessage   |

If so, I will start implementing this.

gep13 commented:
**NOTE:** I added another task above, i.e. a task to go back and update Chocolatey.org, i.e. adding a link to the Gist, updating status, etc.

ferventcoder commented:
Better :)

gep13 commented:
Ok cool, will implement what is documented here, and then work on filling in some of the functionality. :-)
________________________________
From: Rob Reynolds<mailto:notifications@github.com>
Sent: ‎15/‎08/‎2015 15:57
To: chocolatey/package-verifier<mailto:package-verifier@noreply.github.com>
Cc: Gary Ewan Park<mailto:gep13@gep13.co.uk>
Subject: Re: [package-verifier] (WIP) DO NOT MERGE (#5)

Better :)

---
Reply to this email directly or view it on GitHub:
https://github.com/chocolatey/package-verifier/pull/5#issuecomment-131390493

gep13 commented:
@ferventcoder I have made some modifications to how the Tasks and Messages interact, and I believe that they now work as documented.

![image](https://cloud.githubusercontent.com/assets/1271146/9295181/88130aaa-445c-11e5-8603-70c5cad124ab.png)

gep13 commented:
@ferventcoder with regard to Gist creation, we are going to need to use a set of credentials to do this.  I am assuming that we are going to use the choco user to do this, but where do you want those credentials stored?  Configuration File? Environment Variable? Somewhere else?

gep13 commented:
**NOTE:** I will be condensing some of these commits to make them more succinct.  Very much a work in progress :smile_cat:

ferventcoder commented:
Config file - dummies in the one checked in for now.

ferventcoder commented on src/chocolatey.package.verifier/infrastructure.app/tasks/CheckForSubmittedPackagesTask.cs:
```diff
@@ -56,19 +57,16 @@ private void timer_elapsed(object sender, ElapsedEventArgs e)
         {
             this.Log().Info(() => "Checking for submitted packages to test.");
 
-            _timer.Stop();
```
Keep this one

gep13 replied:
Just for the sake of sanity, the stopping and starting the timer while the task is executing is simply to prevent multiple instances of the task?

ferventcoder replied:
Yes, its part of ensuring that we don't duplicate effort accidentally

ferventcoder replied:
It also builds a nice skew over time if there are multiple services running.

ferventcoder commented on src/chocolatey.package.verifier/infrastructure.app/tasks/CheckForSubmittedPackagesTask.cs:
```diff
@@ -56,19 +57,16 @@ private void timer_elapsed(object sender, ElapsedEventArgs e)
+                new PackageSubmittedMessage(
+                    package.Title, package.Version));
 
             _timer.Stop();
```
Looks like I didn't set this one up right (should be `.Start()`

ferventcoder commented on src/chocolatey.package.verifier/infrastructure.app/tasks/ModeratePackageTask.cs:
```diff
@@ -0,0 +1,56 @@
+        {
+            this.Log().Info(
+                () => "Moderating Package: {0} Version: {1}"
+                          .format_with(message.PackageName, message.PackageVersion));
```
We need to bump up your wrap.

ferventcoder replied:
PackageId, not PackageName. :)

ferventcoder commented on src/chocolatey.package.verifier/infrastructure.app/tasks/ModeratePackageTask.cs:
```diff
@@ -0,0 +1,56 @@
+    using infrastructure.messaging;
+    using infrastructure.tasks;
+
+    public class ModeratePackageTask : ITask
```
Not a huge fan of this name

ferventcoder replied:
It's not moderating a package, it's only testing the package.

ferventcoder commented on src/chocolatey.package.verifier/infrastructure.app/tasks/CheckForSubmittedPackagesTask.cs:
```diff
@@ -56,19 +57,16 @@ private void timer_elapsed(object sender, ElapsedEventArgs e)
-                new CreateGistMessage(
-                    @"C:\temp\install.log", "upgrade log", @"C:\temp\uninstall.log", summary: "passed/failed"));
+                new PackageSubmittedMessage(
+                    package.Title, package.Version));
```
package.Id (not title)

ferventcoder commented on src/chocolatey.package.verifier/infrastructure.app/messaging/PackageSubmittedMessage.cs:
```diff
@@ -0,0 +1,31 @@
+            PackageVersion = packageVersion;
+        }
+
+        public string PackageName { get; private set; }
```
Let's call this `PackageId` instead of Name.

ferventcoder commented:
It looks like you've really taken to this architecture. :+1:

gep13 commented:
>@ferventcoder said...
It looks like you've really taken to this architecture. :+1:

Now that I have flipped how I was approaching the problem, it makes perfect sense :smile:

gep13 commented:
@ferventcoder I have updated PR with all of your comments.

Added basic implementation of Gist creation using Octokit, using credentials stored in app.config file.  This results in this example gist being created:

https://gist.github.com/gep13/2dec65af34c28fc53c50

Let me know what your thoughts are.

ferventcoder commented:
I think this is ready to start squashing down.

ferventcoder commented on src/chocolatey.package.verifier.host/chocolatey.package.verifier.host.csproj:
```diff
@@ -83,15 +87,6 @@
-    </Reference>
-    <Reference Include="Microsoft.Web.DistributedCache">
-      <HintPath>..\packages\Glav.CacheAdapter.2.3\lib\Microsoft.Web.DistributedCache.dll</HintPath>
-    </Reference>
```
Cache adapter probably wasn't necessary anyway.

ferventcoder replied:
Was it removed from the packages.config as well?

ferventcoder replied:
Oh wait, don't see it in there in the first place.

gep13 replied:
I will go through all nuget packages that are in packages.config to see whether we are actually going to use them or not.

ferventcoder commented on src/chocolatey.package.verifier.host/packages.config:
```diff
@@ -1,7 +1,7 @@
 <packages>
   <package id="AutoMapper" version="2.2.1" targetFramework="net40" />
-  <package id="log4net" version="1.2.10" allowedVersions="[1.2.10]" />
+  <package id="log4net" version="1.2.10" allowedVersions="[1.2.10]" requireReinstallation="True" />
```
What does that last bit do?

gep13 replied:
I noticed this as well.  I think this goes hand in hand with some msbuild warnings that I also saw, but never acted on. I think it wants this nuget package to be uninstalled and reinstalled as there is another platform target dll that is better suited after the change of framework to 4.5.2.  I will look into this later.

gep13 replied:
These are the warnings that I was referring to:

![image](https://cloud.githubusercontent.com/assets/1271146/9329174/50421e26-45a7-11e5-9493-67dcb5d17ea1.png)

ferventcoder commented on src/chocolatey.package.verifier/infrastructure.app/messaging/PackageTestedMessage.cs:
```diff
@@ -1,35 +1,39 @@
+{
+    using infrastructure.messaging;
+
+    public class PackageTestedMessage : IMessage
```
PackageTestResultMessage?

gep13 replied:
Okay dokay.

ferventcoder commented on src/chocolatey.package.verifier/infrastructure.app/services/GistService.cs:
```diff
@@ -0,0 +1,58 @@
+            gist.Description = description;
+            gist.Public = isPublic;
+
+            foreach (var filePath in filePaths)
```
It's not files, you will be passed the actual messages in log files.

ferventcoder commented on src/chocolatey.package.verifier/infrastructure.app/services/GistService.cs:
```diff
@@ -0,0 +1,58 @@
+
+            foreach (var filePath in filePaths)
+            {
+                if (!File.Exists(filePath))
```
For FileSystem access, use IFileSystem.

gep13 replied:
Noted.  Since the InstallationLog variable is set to be the full string, and not the path to a file, I don't think I will need the filesystem access anyway.

ferventcoder commented on src/chocolatey.package.verifier/infrastructure.app/services/GistService.cs:
```diff
@@ -0,0 +1,58 @@
+            {
+                if (!File.Exists(filePath))
+                {
+                    // TODO: How do you want to handle exceptions?
```
Error handling in a service is handled by Logging an error message.

`this.Log.Error(...)` - the errors will go out over email messages.

gep13 replied:
So log, but don't throw.  Simply continue with next execution?

ferventcoder replied:
This may bring in the circuit breaker pattern as well so we don't send an email hundreds of times - it tries a few times when it errors and increases wait between tries. Once it hits a certain number of tries it "breaks" the circuit, or shuts off the functionality until someone intervenes.

ferventcoder commented on src/chocolatey.package.verifier/infrastructure.app/services/GistService.cs:
```diff
@@ -0,0 +1,58 @@
+
+        private GitHubClient CreateGitHubClient()
+        {
+            // TODO: What sort of error handling do we want around this?  Can we assume that these values will be correctly set?
```
See the above.

ferventcoder commented on src/chocolatey.package.verifier/infrastructure.app/tasks/CheckForSubmittedPackagesTask.cs:
```diff
@@ -60,17 +61,16 @@ private void timer_elapsed(object sender, ElapsedEventArgs e)
 
             var service = new FeedContext_x0060_1(new Uri("http://chocolatey.org/api/v2/submitted/"));
```
Is there a way the FeedContext can have a better name?

gep13 replied:
Off the top of my head, I don't think so.  I think I am right in saying that this name comes from chocolatey.org, so we could look to change it there, but I am not sure how much benefit that would give.  We could manually update the reference.cs file to make it more readable, but if we re-generate the reference then we would have to go back and change it again.

gep13 replied:
@ferventcoder are you ok leaving this as is?

ferventcoder replied:
Yepo

ferventcoder commented on src/chocolatey.package.verifier/infrastructure.app/messaging/PackageTestedMessage.cs:
```diff
@@ -1,35 +1,39 @@
+
+        public string PackageId { get; private set; }
+        public string PackageVersion { get; private set; }
+        public string InstallationLog { get; private set; }
```
These logs are not pointers to files, they are the actual messages.

gep13 replied:
Gotcha! I will change the implementation to reflect this.

ferventcoder commented on src/chocolatey.package.verifier/infrastructure.app/messaging/WebsiteUpdatedMessage.cs:
```diff
@@ -0,0 +1,23 @@
+{
+    using infrastructure.messaging;
+
+    public class WebsiteUpdatedMessage : IMessage
```
I would like to avoid past tense if possible.

gep13 commented:
>@ferventcoder said...
I think this is ready to start squashing down.

Cool, will do :smile_cat:

gep13 commented:
@ferventcoder I have started squashing the commits, and I can get so far, but then I run into some issues when I want to put some commits together.  For example, ideally I would like "Started implementing a GistService" and "Implemented GistService" to be the same commit.  I have tried re-ordering to the commits to allow a fixup, but that results in a merge conflict.  Is there a "best way" approach to fixing this, or would it just be a case of resolving the merge conflict and then squashing?

ferventcoder commented:
A lot of times it's a merge conflict -how bad is it?

gep13 commented:
It didn't seem horrible, I think I would have been able to fix it. I simply backed out of the rebase with ```git rebase --abort``` and moved on.

gep13 commented:
@ferventcoder ok, I have squashed down the commits as far as I think makes sense, let me know if you wanted them done some more.

I have read through a blog post talking about Lokad (http://www.tobinharris.com/past/2009/1/26/net-circuit-breakers/) but I am not entirely sure how you would like it implemented.  Do you want a try/catch in the Gist Service, which will throw a CreateGistException (custom class that we create), which we can then put into a policy in the CreateGistTask?  Or am I off on the wrong tangent?  Do we want a retry policy, or a circuit breaker?  If the latter, I wasn't exactly sure, based on that example, how to increase the time between attempts.  Do you have a sample?

Based on the usage of the PackageTestResultMessage (which I noticed you already had) this would mean that there would be potentially multiple Gists created for the same package version, i.e. one for install, uninstall, upgrade, etc, based on the PackageTestType enumeration.  Is this what you had intended?  I has sort of envisioned a single gist, with all the log files in one place.  What are your thoughts here?

ferventcoder commented:
> Based on the usage of the PackageTestResultMessage (which I noticed you already had) this would mean that there would be potentially multiple Gists created for the same package version, i.e. one for install, uninstall, upgrade, etc, based on the PackageTestType enumeration. Is this what you had intended? I has sort of envisioned a single gist, with all the log files in one place. What are your thoughts here?

Single gist, with multiple files in it. Exactly like your example. :)

ferventcoder commented:
Perhaps remove the test type enumeration though. I the message I had was based on each piece. Which won't be necessary given our current train of thought.

ferventcoder commented:
> I have read through a blog post talking about Lokad (http://www.tobinharris.com/past/2009/1/26/net-circuit-breakers/) but I am not entirely sure how you would like it implemented. Do you want a try/catch in the Gist Service, which will throw a CreateGistException (custom class that we create), which we can then put into a policy in the CreateGistTask? Or am I off on the wrong tangent? Do we want a retry policy, or a circuit breaker? If the latter, I wasn't exactly sure, based on that example, how to increase the time between attempts. Do you have a sample?


I think we can leave ToDos for now and get this all merged in and pick up from there.

gep13 commented:
@ferventcoder okay, rebased again, with hopefully final changes for now.

Example of Gist created by current code can be seen here:

https://gist.github.com/gep13/2e56c55e1bc0053df415

ferventcoder commented:
So if this is no longer `(WIP) DO NOT MERGE` - you might change the title ;)

gep13 commented:
>@ferventcoder said...
So if this is no longer (WIP) DO NOT MERGE - you might change the title ;)

Good point :stuck_out_tongue:

gep13 commented:
@ferventcoder so what's next boss?
